backend/controller/sql/queries.sql

```diff
@@ -673,6 +677,80 @@ VALUES (
   sqlc.arg('payload')
 );
 
+-- name: GetSubscriptionsNeedingUpdate :many
+SELECT
+  subs.key::subscription_key as key,
+  topic_events.key as cursor,
+  topics.key::topic_key as topic,
+  subs.name
+FROM topic_subscriptions subs
+LEFT JOIN topics ON subs.topic_id = topics.id
+LEFT JOIN topic_events ON subs.cursor = topic_events.id
+WHERE subs.cursor IS DISTINCT FROM topics.head
+  AND subs.state = 'idle';
+
+-- name: GetNextEventForSubscription :one
+WITH cursor AS (
+  SELECT
+    created_at,
+    id
+  FROM topic_events
+  WHERE "key" = sqlc.narg('cursor')::topic_event_key
+)
+SELECT events."key" as event,
+        events.payload
+FROM topics
+LEFT JOIN topic_events as events ON events.topic_id = topics.id
+WHERE topics.key = sqlc.arg('topic')::topic_key
+  AND (events.created_at, events.id) > (SELECT COALESCE(MAX(cursor.created_at), '1900-01-01'), COALESCE(MAX(cursor.id), 0) FROM cursor)
+ORDER BY events.created_at, events.id
+LIMIT 1;
+
+-- name: LockSubscriptionAndGetSink :one
+WITH subscriber AS (
+  -- choose a random subscriber to execute the event
+  SELECT
+    subscribers.sink as sink
+  FROM topic_subscribers as subscribers
+  JOIN deployments ON subscribers.deployment_id = deployments.id
```

Comment:
I was thinking about cases where the subscription was shared across modules, but that's not what we're doing anymore so the bulk of the cases I was worried about are fine.

But there's still some weirdness, like this one:
- Add a module with a topic, a subscription, and one subscriber to that subscription.
- Update the module, now with another subscriber to that subscription
In the db we have:
- 2 deployments, one with `min_replicas = 0`, another with `min_replicas = 1`
- 3 subscribers: 1 for new sink, 2 for the sink that originally subscribed (1 for the original deployment, 1 for the new deployment)
If we don't check for min_replicas, the first subscriber would get roughly twice the amount of events

Another case:
- Same as above but instead of added a subscriber, we remove a subscriber. Actually let's remove the sink completely
- The random choosing of a subscriber won't know that the subscriber is for an old deployment, and it'll start making async calls for a sink that doesnt exist anymore

Reply:
Cron jobs also check for min_replicas as well btw

Reply:
I don't think either cron jobs or this should use deployments. As you illustrate above, it vastly complicates the deployment process vs. just synchronising subscriptions directly with the current schema. That is, the topics and subscriptions that exist are only those that are defined in the current schema. All other subscriptions/topics are dropped.

Reply:
Dropping a topic can be messy I think:
- subscriptions may be behind and want to catch up, even though the subscription will not publish anymore

But that could be fine... Maybe subscriptions/subscribers drop immediately, but topics hang around without causing issues.

Added an issue for the epic: #1660

Reply:
It's fine if it's messy - we'll be relying on manual intervention for now. Same as with Verb upgrades with breaking changes, or breaking changes to topic data types. We don't have the systems in place to do it safely, so it's up to the developers to do the right thing for now.